internal/gitinterface/repository.go

```diff
@@ -53,123 +54,79 @@ func LoadRepository() (*Repository, error) {
 
-// executeGitCommandWithEnv is a helper to execute the specified command in the
-// repository after setting the provided environment variables. Note that the
-// command inherits os.Environ() first. This helper explicitly sets the GIT_DIR.
-func (r *Repository) executeGitCommandWithEnv(env []string, args ...string) (io.Reader, io.Reader, error) {
-	args = append([]string{"--git-dir", r.gitDirPath}, args...)
-	return r.executeGitCommandDirectWithEnv(env, args...)
+// executor initializes a new executor instance to run a Git command with the
+// specified arguments.
+func (r *Repository) executor(args ...string) *executor {
+	return &executor{r: r, args: args, env: os.Environ()}
 }
 
-// executeGitCommandDirectWithEnv is a helper to execute the specified command
-// in the repository after setting the provided environment variables. Note that
-// the command inherits os.Environ() first. This helper executes in the current
-// directory without specifying the GIT_DIR explicitly.
-func (r *Repository) executeGitCommandDirectWithEnv(env []string, args ...string) (io.Reader, io.Reader, error) {
-	cmd := exec.Command(binary, args...)
-	cmd.Env = append(os.Environ(), env...)
-
-	var (
-		stdOut bytes.Buffer
-		stdErr bytes.Buffer
-	)
+// withEnv adds the specified environment variables. Each environment variable
+// must be specified in the form of `key=value`.
+func (e *executor) withEnv(env ...string) *executor {
+	e.env = append(e.env, env...)
+	return e
+}
 
-	cmd.Stdout = &stdOut
-	cmd.Stderr = &stdErr
+// withGitDir adds the `--git-dir` parameter to the command. It executes only
+// once.
+func (e *executor) withGitDir() *executor {
+	e.setGitDirOnce.Do(func() {
+		e.args = append([]string{"--git-dir", e.r.gitDirPath}, e.args...)
+	})
+	return e
+}
 
-	return &stdOut, &stdErr, cmd.Run()
+// withStdIn sets the contents of stdin to be passed in to the command.
+func (e *executor) withStdIn(stdIn *bytes.Buffer) *executor {
+	e.stdIn = stdIn
+	return e
 }
 
-// executeGitCommandWithStdInString is a helper to execute the specified command
-// in the repository with the specified `stdIn`. It automatically adds the
-// explicit `--git-dir` parameter.  It returns the stdout for successful
-// execution as a string, with leading and trailing spaces removed.
-func (r *Repository) executeGitCommandWithStdInString(stdIn *bytes.Buffer, args ...string) (string, error) {
-	args = append([]string{"--git-dir", r.gitDirPath}, args...)
-	stdOut, stdErr, err := r.executeGitCommandDirectWithStdIn(stdIn, args...)
+// execute runs the constructed Git command and returns the contents of stdout.
+// Leading and trailing spaces and newlines are removed. This function also sets
+// the Git directory. This function should be used almost every time; the only
+// exception is when the Git directory must not be set and/or the output is
+// desired without any processing such as the removal of space characters.
+func (e *executor) execute() (string, error) {
+	e.withGitDir()
```

Comment:
This is by design, and I can update the documentation for executeRaw (to be renamed execute) to reflect this better. It doesn't do git dir addition because we sometimes are running commands for a repo that don't have a git dir yet. Clone is one example, as is `git rev-parse --git-dir` we run when loading a repository to identify the git dir in the first place. Both of those also use executor, so I kept the git dir addition away from it. The `executeString` API is the happy path we want to use 99.99% of the time. The only case where we don't use executeString in a slightly higher level function is ReadBlob because we want the trailing newline and everything exactly as it is.

Reply:
double checked: we use executeRaw directly:
a) in ReadBlob to have the raw output for the blob's contents
b) in LoadRepository to identify the git-dir itself (so we can't set --git-dir)
c) in CloneAndFetchRepository when we don't have a local git-dir yet (so we can't set --git-dir)

I could move it and have .withGitDir() be set only if repository has a gitDir 🤔

Reply:
done!